mail.go

```diff
@@ -0,0 +1,118 @@
+package detka
+
+import (
+	"fmt"
+
+	"net/smtp"
+
+	"time"
+
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/mailgun/mailgun-go"
+	"github.com/pkg/errors"
+	"github.com/thrawn01/args"
+	"github.com/thrawn01/detka/models"
+)
+
+type Mailer interface {
+	Send(*models.Message) error
+}
+
+func NewMailer(parser *args.ArgParser) (Mailer, error) {
+	opts := parser.GetOpts()
+	if opts.String("mail-transport") == "mailgun" {
+		required := []string{"mailgun-domain", "mailgun-api-key", "mailgun-public-key"}
+		if err := opts.Required(required); err != nil {
+			return nil, errors.New(
+				fmt.Sprintf("'%s' option is required when 'mailgun' is transport", err.Error()))
+		}
+		return &Mailgun{
+			parser: parser,
+		}, nil
+	}
+	required := []string{"smtp-user", "smtp-server", "smtp-password"}
+	if err := opts.Required(required); err != nil {
+		return nil, errors.New(
+			fmt.Sprintf("'%s' option required when 'smtp' is transport", err.Error()))
+	}
+	return &Smtp{
+		parser: parser,
+	}, nil
+}
+
+type Mailgun struct {
+	parser *args.ArgParser
+}
+
+func (self *Mailgun) Send(msg *models.Message) error {
+	opts := self.parser.GetOpts()
+
+	mail := mailgun.NewMailgun(
+		opts.String("mailgun-domain"),
+		opts.String("mailgun-api-key"),
+		opts.String("mailgun-public-key"))
+
+	newMessage := mail.NewMessage(
+		msg.From,
+		msg.Subject,
+		msg.Text,
+		msg.To,
+	)
+
+	var err error
+	for i := 0; i < opts.Int("transport-retry"); i++ {
+		var id string
+		_, id, err = mail.Send(newMessage)
+		if err != nil {
+			logrus.WithFields(logrus.Fields{
+				"method": "Send()",
+				"type":   "mailgun",
+			}).Error("sendEmail - ", err.Error())
+			time.Sleep(time.Second)
```

Comment:
Most services don't appreciate it when you slam them with requests, which is why I always sleep when I retry requests. This really should be an exponential backoff sleep with an upper limit.
```go
const (
	delayFactor = 0.6
	maxDelay = 3.0
)

// Given the number of retries, calculate a suitable backoff sleep time
func BackOff(retries int) time.Duration {
	if retries == 0 {
		return time.Second * 1
	}
	backoff := delayFactor
	for backoff < maxDelay && retries > 0 {
		backoff *= delayFactor
		retries--
	}
	return time.Second * backoff
}

// In-between retries, sleep so we don't hammer the endpoint with requests
time.Sleep(utils.BackOff(retries))
```